# How to choose the optimal RNA-Seq library characteristics for alternative splicing analysis 

The code for this manuscript is stored at https://github.com/ZarnackGroup/Ladwig_et_al_2024.

# Abstract

Alternative splicing (AS) is a key layer of regulation in eukaryotic gene expression that is investigated in all areas of life sciences. Differences in AS between conditions can be quantified from transcriptome-wide short-read RNA sequencing (RNA-Seq) data with designated computational tools. However, not all short-read RNA-Seq data are equally suited for AS analysis. Here, we perform an exemplary AS analysis to showcase the impact of the RNA-Seq library characteristics on the obtained results. Using three standard ENCODE datasets with widespread AS changes, we modulate read length, read depth and the number of replicates and compare their influence on the detection, quantification and classification of AS events with the state-of-the-art AS algorithm MAJIQ. We find that longer reads and a higher read depth are the most effective measures to improve the sensitivity and precision of the analysis. 
From our results, we provide a recommendation on how to best choose the short-read RNA-Seq library specifications for an AS analysis.

