# Abstract
RNA splicing and protein degradation systems allow functional adaptation of the proteome in response to changing cellular contexts. However, the regulatory mechanisms connecting these processes remain poorly understood. Here, we show that impaired spliceosome assembly caused by USP39 deficiency leads to a pathogenic splicing profile characterized by the use of cryptic 5â€² splice sites. Importantly, disruptive cryptic variants evade mRNA surveillance pathways and are translated into misfolded proteins. These spurious isoforms disrupt proteostasis causing proteotoxic aggregates, ER stress and CHOP-mediated cell death. In response to impaired splicing, eukaryotic cells enhance ubiquitination and ER-phagy to alleviate the pathogenic translation of proteotoxic exons. Our findings show how cryptic splicing-induced proteotoxicity can be mitigated, and provide insight into the molecular pathogenesis of spliceosome-associated15 diseases, such as retinitis pigmentosa.

* The source code is deposited at https://github.com/ZarnackGroup/Prieto_Garcia_et_al_2023.
